src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -1118,14 +1121,113 @@ private async Task<Checksum> ComputeDependentChecksumAsync(SolutionState solutio
+                        }
+
+                        writer.WriteKey("analyzerReferences");
+
+                        var serializer = solution.Workspace.Services.GetRequiredService<ISerializerService>();
+                        foreach (var reference in projectState.AnalyzerReferences.OrderBy(r => (r.FullPath, r.Display)))
+                        {
+                            writer.WriteObjectStart();
+
+                            writer.WriteKey("fullPath");
+                            Write(reference.FullPath);
+                            writer.WriteKey("display");
+                            Write(reference.Display);
+
+                            writer.Write("checksum");
+                            var checksum = serializer.CreateChecksum(reference, cancellationToken);
+                            Write(checksum.ToString());
+
+                            writer.WriteObjectEnd();
+                        }
+
+                        writer.WriteObjectEnd();
+
+                        void Write(string? value)
+                        {
+                            if (value == null)
+                                writer.WriteNull();
+                            else
+                                writer.Write(value);
+                        }
+                    }
+
+                    projectKeyBuilder.AppendLine(",");
+
+                    var compilerKey = DeterministicKey.GetDeterministicKey(
+                        projectState.CompilationOptions!,
+                        projectState.DocumentStates.States.Values.OrderBy(s => (s.FilePath, s.Name)).SelectAsArray(s => DocumentStateSyntaxTreeKey.Create(s)),
+                        projectState.MetadataReferences.OrderBy(r => r.Display).ToImmutableArray(),
+                        projectState.AdditionalDocumentStates.States.Values.OrderBy(s => (s.FilePath, s.Name)).SelectAsArray(s => s.AdditionalText),
+                        cancellationToken: cancellationToken);
+
+                    projectKeyBuilder.Append(compilerKey);
+                    var fullKey = projectKeyBuilder.ToString();
+                    return Checksum.Create(fullKey);
+                }
             }
 
             #endregion
+
+            private sealed class DocumentStateSyntaxTreeKey : SyntaxTreeKey
+            {
+                private readonly DocumentState _state;
+
+                private DocumentStateSyntaxTreeKey(DocumentState state)
+                {
+                    _state = state;
+                }
+
+                public static SyntaxTreeKey Create(DocumentState state)
+                    => new DocumentStateSyntaxTreeKey(state);
+
+                public override string? FilePath => _state.FilePath;
+
+                public override ParseOptions Options => _state.ParseOptions!;
+
+                public override SourceText GetText(CancellationToken cancellationToken = default)
+                    => _state.GetTextSynchronously(cancellationToken);
+            }
```

Comment:
@jaredpar an example of an IDE derivation of SyntaxTreeKey.  As you can see, this allows us to avoid parsing.